src/Forms/RestoreAction.php

```diff
@@ -0,0 +1,155 @@
+        }
+
+        $message = self::getRestoreMessage($item, $restoredItem, $changedLocation);
+
+        return $message;
+    }
+
+    /**
+     * Returns a message which notifies the user of a successful restoration
+     * and if anything has changed
+     *
+     * @param $record
+     * @return bool
+     */
+    public static function getRestoreMessage($originalItem, $restoredItem, $changedLocation = false)
+    {
+        $restoredID = $restoredItem->Title ?: $restoredItem->ID;
+        $restoredType = strtolower($restoredItem->i18n_singular_name());
+
+        if (method_exists($restoredItem, 'CMSEditLink') &&
+        $restoredItem->CMSEditLink()) {
+            $restoredID = sprintf('<a href="%s">%s</a>', $restoredItem->CMSEditLink(), $restoredID);
+        }
+
+        if ($originalItem->URLSegment !== $restoredItem->URLSegment) {
+            $changedProperty = [
+                'property' => 'URL',
+                'value' => '../' . $restoredItem->URLSegment
+            ];
+        } elseif ($originalItem->Title !== $restoredItem->Title) {
+            $changedProperty = [
+                'property' => 'Name',
+                'value' => $restoredItem->Title
+            ];
+        }
+
+        if ($changedLocation) {
+            $message = [
+                'text' => _t(
+                    'SilverStripe\\Admin\\ArchiveAdmin.RESTORE_CHANGEDLOCATION',
+                    'Restored the {model} "{id}" to the top level as original location cannot be found.',
+                    [
+                        'model' => $restoredType,
+                        'id' => $restoredID
+                    ]
+                ),
+                'type' => 'notice',
+            ];
+        } elseif (isset($changedProperty)) {
+            $message = [
+                'text' => _t(
+                    'SilverStripe\\Admin\\ArchiveAdmin.RESTORE_CHANGEDPROPERTY',
+                    'A {model} already exists with the same {property}. "{id}" has been restored with a new {property} ({value}).',
+                    [
+                        'model' => $restoredType,
+                        'property' => $changedProperty['property'],
+                        'id' => $restoredID,
+                        'value' => $changedProperty['value']
+                    ]
+                ),
+                'type' => 'warning',
+            ];
+        } else {
+            $message = [
+                'text' => _t('SilverStripe\\Admin\\ArchiveAdmin.RESTORE_SUCCESS', 'Successfully restored the {model} "{id}"', ['model' => $restoredType, 'id' => $restoredID]),
+                'type' => 'good',
+            ];
+        }
+
+        return $message;
+    }
+
+    /**
+     * Determines whether this record can be restored to it's original location
+     *
+     * @param $record
+     * @return bool
+     */
+    public static function shouldRestoreToRoot($record)
+    {
+        if ($parentID = $record->ParentID) {
```

Comment:
Would this logic only be relevant if the record use the `Hierarchy` extensions. I think it would be better to go off that other wise it might produce weird results for object that manually define a Parent attribute.